stablehlo/transforms/StablehloAggressiveSimplification.cpp

```diff
@@ -0,0 +1,1164 @@
+                                PatternRewriter &rewriter) const override {
+    auto precedingBcast =
+        bcast.getOperand()
+            .getDefiningOp<mlir::stablehlo::DynamicBroadcastInDimOp>();
+    if (!precedingBcast) return failure();
+
+    // Compose broadcast dimensions.
+    SmallVector<int64_t> composition;
+    for (int64_t precedingDim : precedingBcast.getBroadcastDimensions()) {
+      composition.push_back(bcast.getBroadcastDimensions()[precedingDim]);
+    }
+    auto composedBcastDims = rewriter.getDenseI64ArrayAttr(composition);
+
+    rewriter.replaceOpWithNewOp<mlir::stablehlo::DynamicBroadcastInDimOp>(
+        bcast, bcast.getType(), precedingBcast.getOperand(),
+        bcast.getOutputDimensions(), composedBcastDims);
+    return success();
+  }
+};
+
+// If all dimensions are known to be nonexpanding from the attribute, replace
+// the dynamic broadcast with a cast.
+struct DynamicBroadcastInDimAllDimsNonExpanding final
+    : OpRewritePattern<mlir::stablehlo::DynamicBroadcastInDimOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::DynamicBroadcastInDimOp op,
+                                PatternRewriter &rewriter) const override {
+    auto resultType = dyn_cast<RankedTensorType>(op.getResult().getType());
+    if (!resultType) {
+      return rewriter.notifyMatchFailure(op, "requires ranked result type");
+    }
+
+    if (!op.getKnownNonexpandingDimensions() ||
+        static_cast<int64_t>(op.getKnownNonexpandingDimensions()->size()) !=
+            resultType.getRank()) {
+      return rewriter.notifyMatchFailure(
+          op, "known_nonexpanding_dimensions don't cover all output dims");
+    }
+
+    auto cast = rewriter.createOrFold<tensor::CastOp>(op.getLoc(), resultType,
+                                                      op.getOperand());
+    rewriter.replaceOp(op, cast);
+    return success();
+  }
+};
+
+struct NoopReduceOpCanon final : OpRewritePattern<mlir::stablehlo::ReduceOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::ReduceOp op,
+                                PatternRewriter &rewriter) const override {
+    // No dimensions to reduce.
+    if (op.getDimensions().empty()) {
+      rewriter.replaceOp(op, op.getInputs());
+      return success();
+    }
+
+    // If all returned values in the ReduceOp region exists outside the
+    // region, replace the ReduceOp with those values.
+    if (auto retOp = dyn_cast<mlir::stablehlo::ReturnOp>(
+            op.getBody().front().getTerminator())) {
+      Region *retRegion = retOp->getParentRegion();
+      if (llvm::any_of(retOp.getResults(), [retRegion](Value result) {
+            return result.getParentRegion() == retRegion;
+          })) {
+        return failure();
+      }
+
+      rewriter.replaceOp(op, retOp.getResults());
+      return success();
+    }
```

Comment:
I have a question regarding this pattern. If reduction is no-op, doesn't returning init value make more sense? I couldn't find what it is supposed to return when reduction is no-op but returning the value from the outside of the region seems weird. Maybe I misunderstand the spec so please correct me if I am wrong.

Reply:
I don't know if I'm understanding this correctly, it's been a long time since I've played stablehlo.
* The pattern here should just do some checking and then some simple transformations.It mainly checks if the result of the `returnOp` is all from the region of the `reduceOp`. In other words, the operands of returnOp should all come from the region of reduceOp.
* If reduction is no-op, doesn't returning init value make more sense?Are you talking about returnOp?
  Since the check here is wrong, then the reduceOp should be wrong and just return failure.